lib/iris/experimental/regrid.py

```diff
@@ -994,14 +1072,78 @@ def _regrid_indices(cells, depth, points):
     # contribution from one or more source points.
     rows = np.nonzero(sum_weights)
 
-    # Calculate the numerator of the weighted mean (M, 1).
-    numerator = sparse_matrix * src_cube.data.reshape(-1, 1)
+    # NOTE: when source points are masked, this 'sum_weights' is possibly
+    # incorrect and needs re-calculating.  Likewise 'rows' may cover target
+    # cells which happen to get no data.  This is dealt with by adjusting as
+    # required in the '__perform' function, below.
+
+    regrid_info = (sparse_matrix, sum_weights, rows, grid_cube)
+    return regrid_info
+
+
+def _regrid_weighted_curvilinear_to_rectilinear__perform(
+        src_cube, regrid_info):
+    """
+    Second (regrid) part of 'regrid_weighted_curvilinear_to_rectilinear'.
 
-    # Calculate the weighted mean payload.
+    Perform the prepared regrid calculation on a single 2d cube.
+
+    """
+    sparse_matrix, sum_weights, rows, grid_cube = regrid_info
+
+    # Calculate the numerator of the weighted mean (M, 1).
+    is_masked = ma.isMaskedArray(src_cube.data)
+    if not is_masked:
+        data = src_cube.data
+    else:
+        # Use raw data array
+        data = src_cube.data.data
+        # Check if there are any masked source points to take account of.
+        is_masked = np.ma.is_masked(src_cube.data)
```

Comment:
This sort of sounds like the same check as in L1095 - I guess the difference is that here you're looking for masked points rather than a masked array? You _could_ simplify this by expecting that if you have a masked array you will always have masked points, but I'm not too fussed if you don't do this.